### PR TITLE
Tidy classical cotransform part of `TransformProgram`

### PR DESCRIPTION
**Pull request by albi3ro**
**Context:**

When trying to improve the organization and structure of our workflow, one major blocker is being able to handle the classical cotransform and argnums component of the transform program.  Right now we can't defer adding in the device transforms and gradient preprocessing till inside `qml.execute`, because we have to set the classical component of the transform program in the qnode.

One option I am considering is storing `qnode`, `args`, and `kwargs` on the program and lazily computing args and the classical component, instead of eagerily calculating the `argnums` and `classical_cotransforms`. This would allow us to setup the full transform program inside `qml.execute`, instead of the `QNode`.  

The main blocker to making this change was understanding what the classical component was actually doing.  While figuring that out, I rewrote some of the code to reduce nesting and reduce the amount going on in each individual component.  Hopefully this makes it a little easier to understand.

**Description of the Change:**

Reduces the levels of nesting in `TransformProgram._set_all_classical_jacobians` to make it a little easier to understand.

**Benefits:**

In the future, it will be easier to move components around without breaking things.

**Possible Drawbacks:**

Maybe an unnecessary waste of time.

**Related GitHub Issues:**

### CODE REVIEW

File: pennylane/transforms/core/transform_program.py [hunk 1]
```diff
@@ -510,17 +511,13 @@ def _set_all_argnums(self, qnode, args, kwargs, argnums):
...
-                    qnode, argnums, TransformProgram(self[0:index]), args, kwargs
-                )
+                params = _jax_argnums_to_tape_trainable(qnode, argnums, self[:index], args, kwargs)
                 argnums_list.append([qml.math.get_trainable_indices(param) for param in params])
             else:
                 argnums_list.append(None)
 
         self._argnums = argnums_list
 
-        qnode.construct(args, kwargs)
```
Comment (dwierichs) [inline pennylane/transforms/core/transform_program.py:522]: Was this simply unnecessary?

File: pennylane/transforms/core/transform_program.py [hunk 2]
```diff
@@ -410,95 +481,25 @@ def prune_dynamic_transform(self, type_to_keep=1):
...
-                multi_tapes = len(tapes) > 1
-                if not multi_tapes:
-                    classical_jacobian = [classical_jacobian]
-                classical_jacobians.append(classical_jacobian)
-            else:
-                classical_jacobians.append(None)
-        self._classical_jacobians = classical_jacobians
-        # Reset the initial tape
-        qnode.construct(args, kwargs)
+        # autograd and tf cant handle pytrees, so need to squeeze batches
```
Comment (dwierichs) [inline pennylane/transforms/core/transform_program.py:496]: Is this comment at the right place? I'm not quite seeing how it comments the code below.

File: pennylane/transforms/core/transform_program.py [hunk 3]
```diff
@@ -25,6 +25,70 @@
...
+    "tf": _tf_jac,
+    "torch": _torch_jac,
+    "jax": _jax_jac,
+    "jax-jit": _jax_jac,
+}
+
+
+# pylint: disable=unused-argument
+def _classical_preprocessing(qnode, program, *args, argnums=None, **kwargs):
+    """Returns the trainable gate parameters for a given QNode input."""
```
Comment (dwierichs) [inline pennylane/transforms/core/transform_program.py:82]: Don't we need to remove the `shots` kwarg any longer? I guess I'm not sure why we had to anyways, but this change seems unrelated to tidying up, so I'm curious :)

File: pennylane/transforms/core/transform_program.py [hunk 4]
```diff
@@ -25,6 +25,70 @@
...
+    import jax
+
+    if argnums is None:
+        if not isinstance(args[0], jax.numpy.ndarray):
+            raise qml.QuantumFunctionError("No trainable parameters.")
+        argnums = 0
+    return jax.jacobian(classical_function, argnums=argnums)(*args, **kwargs)
+
+
+_jac_map = {
```
Comment (dwierichs) [inline pennylane/transforms/core/transform_program.py:69]: So much better this way!

**Review by dwierichs — APPROVED** [review]
I had some minor questions, otherwise looks good to go to me! 
Thanks for taking the time to make this part of the code base more understandable!

File: pennylane/transforms/core/transform_program.py (hunk 1 above)
Comment (albi3ro) [reply pennylane/transforms/core/transform_program.py:522]: Was necessary before recent change to use `make_tape` instead of `construct + qnode.tape`.  Further motivator for why using `make_tape` is a better option here.

File: pennylane/transforms/core/transform_program.py (hunk 3 above)
Comment (albi3ro) [reply pennylane/transforms/core/transform_program.py:82]: `construct_tape` can accept the shots in `kwargs` and will properly put it onto the tape.  Then the program can properly take into consideration the shots.

File: pennylane/transforms/core/transform_program.py (hunk 2 above)
Comment (albi3ro) [reply pennylane/transforms/core/transform_program.py:496]: Maybe "unsqueeze" would be the right term here?

In `_classical_processing`, we squeeze out a singleton dimension if there is only one tape.  Here we have to stick it back in.

I'd love to find a better solution than this squeeze-unsqueeze combination but 🤷 .  Maybe sometime later.

Comment (dwierichs) [reply pennylane/transforms/core/transform_program.py:496]: Gotcha, thanks! Yeah maybe "unsqueeze" would be more comprehensive

File: pennylane/transforms/core/transform_program.py (hunk 3 above)
Comment (dwierichs) [reply pennylane/transforms/core/transform_program.py:82]: Nice, so is this an improvement, or maybe even a bug fix? Sorry about nagging, but do we want to test this or anything? :sweat_smile:

Comment (albi3ro) [reply pennylane/transforms/core/transform_program.py:82]: Potentially an improvement. We would see a difference if the transform had different classical preprocessing based on whether or not there's shots, but I'm not sure we have any of those.

**Review by andrijapau — APPROVED** [review]
LGTM! 🚀 😄
